File: Tarea7.java

import java.util.ArrayList;
import java.util.Scanner;

public class Tarea7 {
    static Banco banco = new Banco();
    static Scanner leer = new Scanner(System.in);

    public static void main(String[] args) {
        int opcion;
        do {
            System.out.println("----- Menú -----");
            System.out.println("1. Crear cuenta para un empleado");
            System.out.println("2. Acceder a la cuenta de un empleado");
            System.out.println("3. Mostrar información de todos los usuarios");
            System.out.println("4. Mostrar información de un usuario en específico");
            System.out.println("5. Mostrar todas las cuentas del banco");
            System.out.println("6. Mostrar información de una cuenta en específico");
            System.out.println("7. Salir");
            System.out.print("Seleccione una opción: ");
            opcion = leer.nextInt();
            leer.nextLine();

            switch (opcion) {
                case 1:
                    crearCuentaParaEmpleado();
                    break;
                case 2:
                    accederCuentaEmpleado();
                    break;
                case 3:
                    mostrarInformacionTodosUsuarios();
                    break;
                case 4:
                    mostrarInformacionUsuarioEspecifico();
                    break;
                case 5:
                    banco.mostrarTodasLasCuentas();
                    break;
                case 6:
                    mostrarInformacionCuentaEspecifica();
                    break;
                case 7:
                    break;
                default:
                    System.out.println("Seleccione una opción válida.");
            }
        } while (opcion != 7);

        leer.close();
    }

    public static void crearCuentaParaEmpleado() {
        System.out.print("Ingrese el nombre del empleado: ");
        String nombre = leer.nextLine();

        Empleado empleado = banco.buscarEmpleado(nombre);
        if (empleado == null) {
            empleado = new Empleado(nombre);
            banco.agregarEmpleado(empleado);
        }

        System.out.print("Ingrese el tipo de cuenta (A, B o C): ");
        char tipoCuenta;
        do {
            tipoCuenta = leer.nextLine().toUpperCase().charAt(0);
            if (tipoCuenta != 'A' && tipoCuenta != 'B' && tipoCuenta != 'C')
                System.out.print("Ingrese A, B o C: ");
        } while (tipoCuenta != 'A' && tipoCuenta != 'B' && tipoCuenta != 'C');

        System.out.print("Ingrese el número de cuenta: ");
        long numeroCuenta = leer.nextLong();
        leer.nextLine();

        CuentaBancaria cuenta = new CuentaBancaria(numeroCuenta, tipoCuenta);
        empleado.agregarCuenta(cuenta);

        System.out.println("Cuenta creada para el empleado " + nombre);
    }

    public static void accederCuentaEmpleado() {
        System.out.print("Ingrese el nombre del empleado: ");
        String nombreEmpleado = leer.nextLine();

        Empleado empleado = banco.buscarEmpleado(nombreEmpleado);
        if (empleado == null) {
            System.out.println("No se encontró ningún empleado con ese nombre.");
            return;
        }

        int opcion;
        do {
            System.out.println("----- Menú de Cuenta -----");
            System.out.println("1. Agregar dinero a una cuenta");
            System.out.println("2. Retirar dinero de una cuenta");
            System.out.println("3. Volver al menú principal");
            System.out.print("Seleccione una opción: ");
            opcion = leer.nextInt();
            leer.nextLine();

            switch (opcion) {
                case 1:
                    agregarDinero(empleado);
                    break;
                case 2:
                    retirarDinero(empleado);
                    break;
                case 3:
                    System.out.println("Volviendo al menú principal...");
                    break;
                default:
                    System.out.println("Opción no válida. Por favor, seleccione una opción válida.");
            }
        } while (opcion != 3);
    }

    public static void agregarDinero(Empleado empleado) {
        System.out.print("Ingrese el número de cuenta: ");
        long numeroCuenta = leer.nextLong();
        leer.nextLine();

        CuentaBancaria cuenta = empleado.buscarCuenta(numeroCuenta);
        if (cuenta == null) {
            System.out.println("No se encontró ninguna cuenta con ese número.");
            return;
        }

        System.out.print("Ingrese la cantidad a depositar (positiva): ");
        double cantidad = leer.nextDouble();
        leer.nextLine();

        cuenta.depositar(cantidad);
    }

    public static void retirarDinero(Empleado empleado) {
        System.out.print("Ingrese el número de cuenta: ");
        long numeroCuenta = leer.nextLong();
        leer.nextLine();

        CuentaBancaria cuenta = empleado.buscarCuenta(numeroCuenta);
        if (cuenta == null) {
            System.out.println("No se encontró ninguna cuenta con ese número.");
            return;
        }

        System.out.print("Ingrese la cantidad a retirar: ");
        double cantidad = leer.nextDouble();
        leer.nextLine();

        cuenta.retirar(cantidad);
    }

    public static void mostrarInformacionTodosUsuarios() {
        banco.mostrarInformacionTodosEmpleados();
    }

    public static void mostrarInformacionUsuarioEspecifico() {
        System.out.print("Ingrese el nombre del empleado: ");
        String nombreEmpleado = leer.nextLine();

        Empleado empleado = banco.buscarEmpleado(nombreEmpleado);
        if (empleado == null) {
            System.out.println("No se encontró ningún empleado con ese nombre.");
        } else {
            empleado.mostrarInformacion();
        }
    }

    public static void mostrarInformacionCuentaEspecifica() {
        System.out.print("Ingrese el número de cuenta: ");
        long numeroCuenta = leer.nextLong();
        leer.nextLine();

        CuentaBancaria cuenta = banco.buscarCuenta(numeroCuenta);
        if (cuenta == null) {
            System.out.println("No hay cuenta con ese numero");
        } else {
            cuenta.mostrarInformacion();
        }
    }

    public static class Empleado {
        private String nombre;
        private ArrayList<CuentaBancaria> cuentas;

        public Empleado(String nombre) {
            this.nombre = nombre;
            this.cuentas = new ArrayList<>();
        }

        public void agregarCuenta(CuentaBancaria cuenta) {
            cuentas.add(cuenta);
        }

        public String getNombre() {
            return nombre;
        }

        public CuentaBancaria buscarCuenta(long numeroCuenta) {
            for (CuentaBancaria cuenta : cuentas) {
                if (cuenta.getNumeroCuenta() == numeroCuenta) {
                    return cuenta;
                }
            }
            return null;
        }

        public void mostrarInformacion() {
            System.out.println("Nombre del empleado: " + nombre);
            System.out.println("Cuentas:");
            for (CuentaBancaria cuenta : cuentas) {
                cuenta.mostrarInformacion();
            }
        }
    }

    public static class CuentaBancaria {
        private long numeroCuenta;
        private char tipoCuenta;
        private double cantidad;

        public CuentaBancaria(long numeroCuenta, char tipoCuenta) {
            this.numeroCuenta = numeroCuenta;
            this.tipoCuenta = tipoCuenta;
            this.cantidad = 0; 
        }

        public long getNumeroCuenta() {
            return numeroCuenta;
        }

        public double getCantidad() {
            return cantidad;
        }

        public void setCantidad(double cantidad) {
            this.cantidad = cantidad;
        }

        public char getTipoCuenta() {
            return tipoCuenta;
        }

        public void depositar(double cantidad) {
            this.cantidad += cantidad;
            System.out.println("Se realizo el deposito. Nuevo saldo: " + this.cantidad);
        }

        public void retirar(double cantidad) {
            if (this.cantidad - cantidad >= obtenerSaldoMinimo()) {
                this.cantidad -= cantidad;
                System.out.println("Se realizo el retiro Nuevo saldo: " + this.cantidad);
            } else {
                System.out.println("Saldo insuficiente");
            }
        }

        private double obtenerSaldoMinimo() {
            switch (tipoCuenta) {
                case 'A':
                    return 1000;
                case 'B':
                    return 5000;
                case 'C':
                    return 10000;
                default:
                    return 0;
            }
        }

        public void mostrarInformacion() {
            System.out.println("Número de cuenta: " + numeroCuenta);
            System.out.println("Tipo de cuenta: " + tipoCuenta);
            System.out.println("Saldo: " + cantidad);
        }
    }

    public static class Banco {
        private ArrayList<Empleado> empleados;

        public Banco() {
            this.empleados = new ArrayList<>();
        }

        public void agregarEmpleado(Empleado empleado) {
            empleados.add(empleado);
        }

        public Empleado buscarEmpleado(String nombre) {
            for (Empleado empleado : empleados) {
                if (empleado.getNombre().equals(nombre)) {
                    return empleado;
                }
            }
            return null;
        }

        public CuentaBancaria buscarCuenta(long numeroCuenta) {
            for (Empleado empleado : empleados) {
                CuentaBancaria cuenta = empleado.buscarCuenta(numeroCuenta);
                if (cuenta != null) {
                    return cuenta;
                }
            }
            return null;
        }

        public void mostrarInformacionTodosEmpleados() {
            for (Empleado empleado : empleados) {
                empleado.mostrarInformacion();
                System.out.println();
            }
        }

        public void mostrarTodasLasCuentas() {
            for (Empleado empleado : empleados) {
                for (CuentaBancaria cuenta : empleado.cuentas) {
                    cuenta.mostrarInformacion();
                }
            }
        }
    }
}

File: Tarea7.py
class Empleado:
    def __init__(self, nombre):
        self.nombre = nombre
        self.cuentas = []

    def agregar_cuenta(self, numero_cuenta, tipo_cuenta):
        self.cuentas.append(CuentaBancaria(numero_cuenta, tipo_cuenta))

    def obtener_nombre(self):
        return self.nombre

    def buscar_cuenta(self, numero_cuenta):
        for cuenta in self.cuentas:
            if cuenta.obtener_numero_cuenta() == numero_cuenta:
                return cuenta
        return None

    def mostrar_informacion(self):
        print("Nombre del empleado:", self.nombre)
        print("Cuentas:")
        for cuenta in self.cuentas:
            cuenta.mostrar_informacion()
            print()


class CuentaBancaria:
    def __init__(self, numero_cuenta, tipo_cuenta):
        self.numero_cuenta = numero_cuenta
        self.tipo_cuenta = tipo_cuenta
        self.saldo = 0

    def obtener_numero_cuenta(self):
        return self.numero_cuenta

    def obtener_tipo_cuenta(self):
        return self.tipo_cuenta

    def depositar(self, cantidad):
        self.saldo += cantidad
        print("Se realizó el depósito. Nuevo saldo:", self.saldo)

    def retirar(self, cantidad):
        if self.saldo - cantidad >= self.obtener_saldo_minimo():
            self.saldo -= cantidad
            print("Se realizó el retiro. Nuevo saldo:", self.saldo)
        else:
            print("Saldo insuficiente")

    def obtener_saldo_minimo(self):
        if self.tipo_cuenta == 'A':
            return 1000
        elif self.tipo_cuenta == 'B':
            return 5000
        elif self.tipo_cuenta == 'C':
            return 10000
        else:
            return 0

    def mostrar_informacion(self):
        print("Número de cuenta:", self.numero_cuenta)
        print("Tipo de cuenta:", self.tipo_cuenta)
        print("Saldo:", self.saldo)


class Banco:
    def __init__(self):
        self.empleados = []

    def agregar_empleado(self, nombre):
        self.empleados.append(Empleado(nombre))

    def buscar_empleado(self, nombre):
        for empleado in self.empleados:
            if empleado.obtener_nombre() == nombre:
                return empleado
        return None

    def buscar_cuenta(self, numero_cuenta):
        for empleado in self.empleados:
            cuenta = empleado.buscar_cuenta(numero_cuenta)
            if cuenta:
                return cuenta
        return None

    def mostrar_informacion_todos_empleados(self):
        for empleado in self.empleados:
            empleado.mostrar_informacion()
            print()

    def mostrar_todas_las_cuentas(self):
        for empleado in self.empleados:
            for cuenta in empleado.cuentas:
                cuenta.mostrar_informacion()


banco = Banco()

while True:
    print("----- Menú -----")
    print("1. Crear cuenta para un empleado")
    print("2. Acceder a la cuenta de un empleado")
    print("3. Mostrar información de todos los usuarios")
    print("4. Mostrar información de un usuario en específico")
    print("5. Mostrar todas las cuentas del banco")
    print("6. Mostrar información de una cuenta en específico")
    print("7. Salir")

    opcion = int(input("Seleccione una opción: "))

    if opcion == 1:
        nombre = input("Ingrese el nombre del empleado: ")
        if not banco.buscar_empleado(nombre):
            banco.agregar_empleado(nombre)

        tipo_cuenta = input("Ingrese el tipo de cuenta (A, B o C): ").upper()
        while tipo_cuenta not in ['A', 'B', 'C']:
            tipo_cuenta = input("Ingrese A, B o C: ").upper()

        numero_cuenta = int(input("Ingrese el número de cuenta: "))
        banco.buscar_empleado(nombre).agregar_cuenta(numero_cuenta, tipo_cuenta)

        print("Cuenta creada para el empleado", nombre)

    elif opcion == 2:
        nombre_empleado = input("Ingrese el nombre del empleado: ")
        empleado = banco.buscar_empleado(nombre_empleado)
        if not empleado:
            print("No se encontró ningún empleado con ese nombre.")
        else:
            while True:
                print("----- Menú de Cuenta -----")
                print("1. Agregar dinero a una cuenta")
                print("2. Retirar dinero de una cuenta")
                print("3. Volver al menú principal")

                opcion_cuenta = int(input("Seleccione una opción: "))

                if opcion_cuenta == 1:
                    numero_cuenta = int(input("Ingrese el número de cuenta: "))
                    cuenta = empleado.buscar_cuenta(numero_cuenta)
                    if not cuenta:
                        print("No se encontró ninguna cuenta con ese número.")
                    else:
                        cantidad = float(input("Ingrese la cantidad a depositar (positiva): "))
                        cuenta.depositar(cantidad)

                elif opcion_cuenta == 2:
                    numero_cuenta = int(input("Ingrese el número de cuenta: "))
                    cuenta = empleado.buscar_cuenta(numero_cuenta)
                    if not cuenta:
                        print("No se encontró ninguna cuenta con ese número.")
                    else:
                        cantidad = float(input("Ingrese la cantidad a retirar: "))
                        cuenta.retirar(cantidad)

                elif opcion_cuenta == 3:
                    print("Volviendo al menú principal...")
                    break

                else:
                    print("Opción no válida. Por favor, seleccione una opción válida.")

    elif opcion == 3:
        banco.mostrar_informacion_todos_empleados()

    elif opcion == 4:
        nombre_empleado = input("Ingrese el nombre del empleado: ")
        empleado = banco.buscar_empleado(nombre_empleado)
        if not empleado:
            print("No se encontró ningún empleado con ese nombre.")
        else:
            empleado.mostrar_informacion()

    elif opcion == 5:
        banco.mostrar_todas_las_cuentas()

    elif opcion == 6:
        numero_cuenta = int(input("Ingrese el número de cuenta: "))
        cuenta = banco.buscar_cuenta(numero_cuenta)
        if not cuenta:
            print("No hay cuenta con ese número")
        else:
            cuenta.mostrar_informacion()

    elif opcion == 7:
        break

    else:
        print("Seleccione una opción válida.")
